stub/src/main/java/io/grpc/stub/ServerCalls.java

```diff
@@ -122,24 +129,35 @@ private ServerCalls() {
           ReqT request;
           @Override
           public void onMessage(ReqT request) {
+            if (this.request != null) {
+              // Safe to close the call, because the application has not yet been invoked
```

Comment:
invoke() can still happen after this point. It seems you may need a boolean to ignore any future halfClose.